### PR TITLE
Feature: Add Testground runtime Dockerfile

### PR DESCRIPTION
**Pull request by Kohelbekker**

### CODE REVIEW

File: testground/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+RUN mkdir -p /usr/local/bin
+COPY --from=0 /testground-master/testground /testground
+COPY --from=0 /testground-master/static /static
+COPY --from=0 /testground-master/tmpl /tmpl
+COPY --from=0 /testground-master/.env.toml /root/testground/.env.toml
+ENV PATH="/:/usr/local/bin:${PATH}"
+
+EXPOSE 8042
+
+WORKDIR "/"
```
Comment (vshatravenko) [inline testground/Dockerfile:45]: Do we need this?

File: testground/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ARG GO_VERSION=1.16
+
+# Dynamically select the golang version.
+FROM golang:${GO_VERSION}-buster
+
+RUN apt-get update && apt-get install unzip
+
+RUN wget https://github.com/testground/testground/archive/refs/heads/master.zip && unzip master -d /
+
+#  COPY /go.mod /go.mod
```
Comment (vshatravenko) [inline testground/Dockerfile:15]: Let's remove this if we don't need it

File: testground/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# Download deps.
+RUN cd /testground-master && go mod download
+
+# Now copy the rest of the source and run the build.
+# COPY . /
+
+COPY .env.toml /testground-master
+
+# Testground version
+ARG TG_VERSION
```
Comment (vshatravenko) [inline testground/Dockerfile:26]: Let's put this at the top, the same as GO_VERSION

File: testground/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+FROM golang:${GO_VERSION}-buster
+
+RUN apt-get update && apt-get install unzip
+
+RUN wget https://github.com/testground/testground/archive/refs/heads/master.zip && unzip master -d /
+
+#  COPY /go.mod /go.mod
+
+# Download deps.
+RUN cd /testground-master && go mod download
```
Comment (vshatravenko) [inline testground/Dockerfile:18]: Let's do `WORKDIR /app` instead and use /app in all paths

File: testground/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#::: RUNTIME CONTAINER
+#:::
+
+FROM golang:${GO_VERSION}-buster
+
+RUN mkdir -p /usr/local/bin
+COPY --from=0 /testground-master/testground /testground
+COPY --from=0 /testground-master/static /static
+COPY --from=0 /testground-master/tmpl /tmpl
+COPY --from=0 /testground-master/.env.toml /root/testground/.env.toml
```
Comment (vshatravenko) [inline testground/Dockerfile:40]: Let's just do `COPY /app/testground /app/static ... /`

File: testground/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ENV PATH="/:/usr/local/bin:${PATH}"
```
Comment (vshatravenko) [inline testground/Dockerfile:41]: I'd rather do a specific dir for testground instead of putting `/` into `PATH`

File: testground/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+EXPOSE 8042
+
+WORKDIR "/"
+
+ENTRYPOINT [ "/testground", "--vv", "daemon" ]
```
Comment (vshatravenko) [inline testground/Dockerfile:47]: If it's in the PATH, do we need the leading `/`?

File: testground/README.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Testground
+
+This docker image is used for testing of OpenDAX-v4
```
Comment (vshatravenko) [inline testground/README.md:3]: I'd rather write that this image is used as the runtime for running Testground tests in general

**Review by vshatravenko — CHANGES_REQUESTED** [review]

File: testground/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X github.com/testground/testground/pkg/version.GitCommit=${TG_VERSION}" -o testground
+
+#:::
+#::: RUNTIME CONTAINER
+#:::
+
+FROM golang:${GO_VERSION}-buster
+
+RUN mkdir -p /usr/local/bin && mkdir -p /app
```
Comment (vshatravenko) [inline testground/Dockerfile:33]: Do we need a /usr/local/bin?

Comment (vshatravenko) [reply testground/Dockerfile:33]: Also, we can simply do a `WORKDIR /app`, it will create the dir automatically

**Review by vshatravenko — CHANGES_REQUESTED** [review]

File: testground/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+#:::
+
+# GO_VERSION is the golang version this image will be built against.
+ARG GO_VERSION=1.16
+
+# Dynamically select the golang version.
+FROM golang:${GO_VERSION}-buster
+
+# Testground version
+COPY VERSION /VERSION
```
Comment (vshatravenko) [inline testground/Dockerfile:12]: This is supposed to be passed to `docker build` as a build arg, I think the Ruby logic already does that
You can check existing images/builds for that

File: testground/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+#:::
+#::: RUNTIME CONTAINER
+#:::
+
+FROM golang:${GO_VERSION}-buster
+
+WORKDIR /app
+
+COPY --from=0 /app/testground /app/
+COPY --from=0 /app/.env.toml /root/testground/.env.toml
```
Comment (vshatravenko) [inline testground/Dockerfile:40]: Let's try to have only one COPY statement here

File: testground/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ENV PATH="/app:${PATH}"
+
+EXPOSE 8042
+
+ENTRYPOINT [ "testground", "--vv", "daemon" ]
```
Comment (vshatravenko) [inline testground/Dockerfile:45]: I'm not sure we should have daemon in the entrypoint, testground should be enough

**Review by vshatravenko — CHANGES_REQUESTED** [review]

File: testground/Dockerfile (hunk 12 above)
Comment (Kohelbekker) [reply testground/Dockerfile:45]: But it's Dockerfile for testground daemon, we never use it for any other purpose plus it's how it was created by testground team

File: testground/Dockerfile (hunk 11 above)
Comment (Kohelbekker) [reply testground/Dockerfile:40]: But there are 2 different destinations

**Review by vshatravenko — APPROVED** [review]

**Pull request merged**